File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;
class Solution {
    static Map<String, Integer> map;
    static int[][] giftCountList;
    static int[] giftScore;
    static int[] answerList;
    static int maxCount = 0;
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        int memberCount = friends.length;
        giftCountList = new int[memberCount][memberCount];
        giftScore = new int[memberCount];
        answerList = new int[memberCount];
        
        map = new HashMap<>();
        
        for (int i = 0; i < memberCount; i++) {
            map.put(friends[i], i);
        }
        
        for (String gift: gifts) {
            StringTokenizer st = new StringTokenizer(gift);
            String sender = st.nextToken();
            String receiver = st.nextToken();
            
            giftCountList[map.get(sender)][map.get(receiver)]++;
            giftScore[map.get(sender)]++;
            giftScore[map.get(receiver)]--;
        }
        
        for (int i = 0; i < memberCount; i++) {
            for (int j = i + 1; j < memberCount; j++) {
                int sendPoint = giftCountList[i][j];
                int receivePoint = giftCountList[j][i];
                
                if (sendPoint == receivePoint) {
                    if (giftScore[i] > giftScore[j]) {
                        addCount(i);
                    } else if (giftScore[i] < giftScore[j]) {
                        addCount(j);
                    }
                } else if (sendPoint > receivePoint){
                    addCount(i);
                } else {
                    addCount(j);
                }
            }
        }
        
        return maxCount;
    }
    static void addCount(int idx) {
        answerList[idx]++;
        maxCount = Math.max(maxCount, answerList[idx]);
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.kt
import java.util.*

class Solution {
    fun solution(friends: Array<String>, gifts: Array<String>): Int {
        val giftCount = mutableMapOf<String, Int>()
        val totalCount = mutableMapOf<String, Int>()
        for (name in friends) {
            giftCount.put(name, 0)
            totalCount.put(name, 0)
        }
        
        val history = mutableMapOf<String, Int>()
        for (gift in gifts) {
            val split = gift.split(" ")
            
            giftCount[split[0]] = giftCount[split[0]]!! + 1
            giftCount[split[1]] = giftCount[split[1]]!! - 1
            history[gift] = history.getOrPut(gift) { 0 } + 1
        }
        
        val size = friends.size
        
        for (i in 0 until size) {
            val l = friends[i]
            for (j in i until size) {
                val r = friends[j]
                val lStr = "$l $r"
                val rStr = "$r $l"
                
                val lCount = history.getOrDefault(lStr, 0)
                val rCount = history.getOrDefault(rStr, 0)
                
                val targetName = if (lCount == rCount) {
                    if (giftCount[l] == giftCount[r]) continue
                    else {
                        if (giftCount[l]!! > giftCount[r]!!) l else r
                    }
                } else {
                    if (lCount > rCount) l else r
                }
                totalCount[targetName] = totalCount[targetName]!! + 1
            }
        }
        var max = 0
        for (n in totalCount.values) {
            max = maxOf(max, n)
        }
        
        return max
    }
}